minseo/app.js

```diff
@@ -36,6 +36,53 @@ app.get("/ping", function (req, res, next) {
   res.json({ message: "pong" });
 });
 
+app.post("/posts", async (req, res) => {
+  const { title, content, userId } = req.body;
+
+  await appDataSource.query(
+    `INSERT INTO posts(
+    title,
+    content,
+    user_id
+  )VALUES(?, ?, ?);`,
+    [title, content, userId]
+  );
+  res.status(201).json({ message: "postCreated" });
+});
+
+app.get("/posts", async (req, res) => {
+  await appDataSource.query(
+    `SELECT
+      users.id AS userId,
+      posts.id AS postingId,
+      posts.content AS postingContent
+      FROM users
+      JOIN posts ON users.id = posts.user_id
+      `,
+    (err, rows) => {
+      res.status(200).json(rows);
+    }
+  );
+});
+app.get("/user_posts", async (req, res) => {
+  const { userId } = req.body;
+  await appDataSource.query(
+    `SELECT 
+      users.id AS userId,
+      users.profileImage AS userProfileImage,
+      JSON_ARRAYAGG(JSON_OBJECT(
+        "postingId",posts.id,
+        "postingImage",posts.postingImage,
+        "postingContent",posts.content))
+         AS postings
+      FROM users INNER JOIN posts ON users.id = posts.user_id WHERE users.id = ${userId} GROUP BY users.id
```

Comment:
```suggestion
      FROM users INNER JOIN posts ON users.id = posts.user_id WHERE users.id = ? GROUP BY users.id
```
typeorm의 initialize method를 통해 만들어진 객체의 .query()에는 지금 우리가 작성하고 있는 것처럼 raw SQL문이 들어갈 수 있습니다.

그 안에 작성되는 SQL 문에는 00님이 작성해주신 것과 같이 템플릿 리터럴(`${ }`)을 통해 변수를 넣을 수도, 혹은 제가 suggestion으로 작성드린 SQL 안에 placeholder라고 하는 자리 표시자(`?`)를 지정하여 동적으로 쿼리문이 수행될 수 있도록 값을 넣을 수도 있습니다.

위처럼 두가지 방법 모두 다 raw SQL 안에 입력값을 넣어줄 수 있습니다.

다만 우리는 가급적 placeholder(`?`)를 사용하여 값을 넣도록 지향해야 하는 이유는 보안적인 측면과 관련있습니다.

백엔드는 비단 클라이언트와 데이터베이스 서버 사이의 접점이라고 볼 수 있는 API를 단순 동작하도록 구현하는 것 뿐만이 아닌, 클라이언트와 요청과 응답을 주고 받을 때, 또는 database에서 원하는 data를 추출할 때를 포함한 여러 과정에서의 보안성도 고려해야 합니다.

위 내용과 가장 관련이 있는 악성 공격으로는 SQL Injection 이라는 것이 있고 이는 정상적인 사용자가 아닌 누군가가 SQL문에 중요한 데이터를 추출할 수 있는 값을(ex. 항상 TRUE가 될 수 있는) 의도적으로 삽입 하여 data를 탈취하는 공격이고, 템플릿 리터럴(`${ }`)을 사용하면 이에 취약한 코드가 된다고 볼 수 있습니다.

자세한 동작 원리는 생략하겠지만, `?`를 사용하여 파라미터를 바인딩하게 되면 템플릿 리터럴 방식보다 sql injection에 해서 조금 더 안전하게 raw SQL을 사용할 수 있게 됩니다.

따라서 가능하면 place holder(`?`)를 사용하여 parameter binding을 통해 변수 값을 넣어주는 방식으로 수정 혹은 통일하여 구현해주세요!